Lib/typing.py

```diff
@@ -1831,7 +1831,8 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
         if value is None:
             value = type(None)
         if isinstance(value, str):
-            value = ForwardRef(value)
+            value = ForwardRef(
+                value, is_argument=not isinstance(obj, types.ModuleType))
```

Comment:
Could you please explain why this only applies to module objects? Why not set it to `False` always?

Also, doesn't this mean top-level ClassVars are no longer considered invalid? That doesn't seem right.

Reply:
Yes, you are right. I hoped that tests can catch any regressions in this logic, but there was no test for module-level `ClassVar`.

This problem is more complex than I initially thought. Right now `ForwardRef` only has one boolean flag that changes its behavior: `is_argument`. 

https://github.com/python/cpython/blob/8492b729ae97737d22544f2102559b2b8dd03a03/Lib/typing.py#L166-L168

But we have at least three possible places where different types are treated differently:
1. Module / local level, we cannot use `ClassVar` here. `Final` is fine
2. Class level, we can use both
3. Function / method arguments, we cannot use both `ClassVar` and `Final`

I suggest to add one more flag to `ForwardRef` and `_type_check`: `is_class` argument with `False` as default. This way we will be able to represent all three possible states.

However, I understand that this is big change to the API. `_type_check` is not public, but `ForwardRef` is public. What is the policy for such cases?

In the meantime I will update this PR with a prototype, so you can take a closer look at my suggestion.